calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,380 @@
+                    && dst.attributes.has("go")
+                    && comp_or_non_comb(&src_cell.upgrade())
+                    && comp_or_non_comb(&dst_cell.upgrade())
+                {
+                    self.done_go = Some((
+                        src_cell.upgrade().clone_name(),
+                        dst_cell.upgrade().clone_name(),
+                    ));
+                }
+            }
+            // src_cell's done writes to group's done
+            (ir::PortParent::Cell(src_cell), ir::PortParent::Group(_)) => {
+                if dst.name == "done"
+                    && src.attributes.has("done")
+                    && comp_or_non_comb(&src_cell.upgrade())
+                {
+                    self.last = Some(src_cell.upgrade().borrow().clone_name())
+                }
+            }
+            // If we encounter anything else, then not of interest to us
+            _ => (),
+        }
+    }
+    //Returns whether the given assignment is a go-done assignment
+    //i.e. cell1.go = cell2.done.
+    pub fn is_go_done(asmt: &ir::Assignment) -> bool {
+        let src = asmt.src.borrow();
+        let dst = asmt.dst.borrow();
+        match (&src.parent, &dst.parent) {
+            (ir::PortParent::Cell(_), ir::PortParent::Cell(_)) => {
+                src.attributes.has("done") && dst.attributes.has("go")
+            }
+            _ => false,
+        }
+    }
+
+    //Returns whether the given assignment writes to the go assignment of cell
+    //in the form cell.go = !cell.done? 1'd1.
+    pub fn is_specific_go(asmt: &ir::Assignment, cell: &ir::Id) -> bool {
+        //checks whether guard is cell.done
+        let guard_is_done = |guard: &ir::Guard| -> bool {
+            match guard {
+                ir::Guard::Port(port) => {
+                    port.borrow().attributes.has("done")
+                        && port.borrow().get_parent_name() == cell
+                }
+                _ => false,
+            }
+        };
+
+        //checks whether guard is !cell.done
+        let guard_not_done = |guard: &ir::Guard| -> bool {
+            match guard {
+                ir::Guard::Not(g) => guard_is_done(&*g),
+                _ => false,
+            }
+        };
+
+        let dst = asmt.dst.borrow();
+        // checks cell.go =
+        dst.get_parent_name() == cell  && dst.attributes.has("go")
+        // checks !cell.done ?
+        && guard_not_done(&*asmt.guard)
+        // checks 1'd1
+        && asmt.src.borrow().is_constant(1, 1)
+    }
+
+    /// Returns group with made using builder with prefix. The assignments are
+    /// asmts, plus a write to groups's done, based on done_src and done_guard.
+    fn make_group(
+        done_src: ir::RRC<ir::Port>,
+        done_guard: ir::Guard,
+        asmts: Vec<ir::Assignment>,
+        builder: &mut ir::Builder,
+        prefix: String,
+    ) -> ir::RRC<ir::Group> {
+        let group = builder.add_group(prefix);
```

Comment:
Can you open an issue saying that `add_group` should always take a done condition?